libevm/triedb/firewood/firewood.go

```diff
@@ -0,0 +1,518 @@
+			},
+		},
+	}, nil
+}
+
+func validatePath(dir string) error {
+	if dir == "" {
+		return errors.New("chain data directory must be set")
+	}
+
+	// Check that the directory exists
+	switch info, err := os.Stat(dir); {
+	case os.IsNotExist(err):
+		log.Info("Database directory not found, creating", "path", dir)
+		if err := os.MkdirAll(dir, 0o755); err != nil {
+			return fmt.Errorf("error creating database directory: %w", err)
+		}
+		return nil
+	case err != nil:
+		return fmt.Errorf("error checking database directory: %w", err)
+	case !info.IsDir():
+		return fmt.Errorf("database directory path is not a directory: %s", dir)
+	}
+
+	return nil
+}
+
+// Scheme returns the scheme of the database.
+// This is only used in some API calls
+// and in StateDB to avoid iterating through deleted storage tries.
+// WARNING: If anything from go-ethereum uses this,
+// it must be overwritten to use something like:
+// `_, ok := db.(*Database); if !ok { return "" }`
+// to recognize the Firewood database.
+func (*Database) Scheme() string {
+	return rawdb.HashScheme
+}
+
+// Initialized checks whether a non-empty genesis block has been written.
+func (db *Database) Initialized(common.Hash) bool {
+	rootBytes, err := db.Firewood.Root()
+	if err != nil {
+		log.Error("firewood: error getting current root", "error", err)
+		return false
+	}
+	root := common.BytesToHash(rootBytes)
+	// If the current root isn't empty, then unless the database is empty, we have a genesis block recorded.
+	return root != types.EmptyRootHash
+}
+
+// Size returns the storage size of diff layer nodes above the persistent disk
+// layer and the dirty nodes buffered within the disk layer
+// Only used for metrics and Commit intervals in APIs.
+// This will be implemented in the firewood database eventually.
+// Currently, Firewood stores all revisions in disk and proposals in memory.
+func (*Database) Size() (common.StorageSize, common.StorageSize) {
+	return 0, 0
+}
+
+// This isn't called anywhere in coreth
+func (*Database) Reference(common.Hash, common.Hash) {}
+
+// Dereference drops a proposal from the database.
+// This function is no-op because unused proposals are dereferenced when no longer valid.
+// We cannot dereference at this call. Consider the following case:
+// Chain 1 has root A and root C
+// Chain 2 has root B and root C
+// We commit root A, and immediately dereference root B and its child.
+// Root C is Rejected, (which is intended to be 2C) but there's now only one record of root C in the proposal map.
+// Thus, we recognize the single root C as the only proposal, and dereference it.
+func (*Database) Dereference(common.Hash) {}
+
+// Firewood does not support this.
+func (*Database) Cap(common.StorageSize) error {
+	return nil
+}
+
+func (db *Database) Close() error {
+	db.proposalLock.Lock()
+	defer db.proposalLock.Unlock()
+
+	// All remaining proposals must be dereferenced.
+	db.possibleProposals = nil
+	db.proposalMap = nil
+	db.proposalTree = nil
```

Comment:
Should we not just explicitly drop them? The finalizer semantics were intended as a fail-safe just in case we "lose" one in the `libevm` plumbing, but in this case it's all under our control.